runners/flink/src/test/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunctionTest.java

```diff
@@ -0,0 +1,245 @@
+
+    @SuppressWarnings("unchecked")
+    RemoteBundle<Integer> bundle = Mockito.mock(RemoteBundle.class);
+    when(stageBundleFactory.<Integer>getBundle(any(), any())).thenReturn(bundle);
+
+    @SuppressWarnings("unchecked")
+    FnDataReceiver<WindowedValue<Integer>> receiver = Mockito.mock(FnDataReceiver.class);
+    when(bundle.getInputReceiver()).thenReturn(receiver);
+
+    Exception expected = new Exception();
+    doThrow(expected).when(bundle).close();
+    thrown.expect(is(expected));
+    function.mapPartition(Collections.emptyList(), collector);
+  }
+
+  @Test
+  public void checksForRuntimeContextChanges() throws Exception {
+    FlinkExecutableStageFunction<Integer> function = getFunction(Collections.emptyMap());
+    function.open(new Configuration());
+    // Change runtime context.
+    function.setRuntimeContext(Mockito.mock(RuntimeContext.class));
+    thrown.expect(Matchers.instanceOf(IllegalStateException.class));
+    function.mapPartition(Collections.emptyList(), collector);
+  }
+
+  @Test
+  public void inputsAreSentInOrder() throws Exception {
+    FlinkExecutableStageFunction<Integer> function = getFunction(Collections.emptyMap());
+    function.open(new Configuration());
+
+    @SuppressWarnings("unchecked")
+    RemoteBundle<Integer> bundle = Mockito.mock(RemoteBundle.class);
+    when(stageBundleFactory.<Integer>getBundle(any(), any())).thenReturn(bundle);
+
+    @SuppressWarnings("unchecked")
+    FnDataReceiver<WindowedValue<Integer>> receiver = Mockito.mock(FnDataReceiver.class);
+    when(bundle.getInputReceiver()).thenReturn(receiver);
+
+    WindowedValue<Integer> one = WindowedValue.valueInGlobalWindow(1);
+    WindowedValue<Integer> two = WindowedValue.valueInGlobalWindow(2);
+    WindowedValue<Integer> three = WindowedValue.valueInGlobalWindow(3);
+    function.mapPartition(Arrays.asList(one, two, three), collector);
+
+    InOrder order = Mockito.inOrder(receiver);
+    order.verify(receiver).accept(one);
+    order.verify(receiver).accept(two);
+    order.verify(receiver).accept(three);
+  }
+
+  @Test
+  public void outputsAreTaggedCorrectly() throws Exception {
+    WindowedValue<Integer> three = WindowedValue.valueInGlobalWindow(3);
+    WindowedValue<Integer> four = WindowedValue.valueInGlobalWindow(4);
+    WindowedValue<Integer> five = WindowedValue.valueInGlobalWindow(5);
+    Map<String, Integer> outputTagMap =
+        ImmutableMap.of(
+            "one", 1,
+            "two", 2,
+            "three", 3);
+    FlinkExecutableStageFunction<Integer> function = getFunction(outputTagMap);
+    function.open(new Configuration());
+
+    CompletableFuture<OutputReceiverFactory> receiverFactoryFuture = new CompletableFuture<>();
+
+    @SuppressWarnings("unchecked")
+    RemoteBundle<Integer> bundle = Mockito.mock(RemoteBundle.class);
+    // Capture the real receiver factory when a bundle is requested.
+    when(stageBundleFactory.<Integer>getBundle(any(), any()))
+        .thenAnswer(
```

Comment:
Not sure if mocks are the best way to handle this, maybe anonymous classes would be better because you would't be losing static typing.

Reply:
I agree. However, it's painful to write this full set of nested factories uniquely for each test case. I think this will be better once we only have a single Context that handles all the auxiliary details, but I wan to figure out exactly what that looks like.